Prometheus: Add off switch for metric/label name lookup

s-h-a-d-o-w opened the pull request:
**What this PR does / why we need it**:
This will help users with amounts of metric name data that is too much for a browser to handle.
Query editor autocomplete will be disabled and metrics chooser hidden, since obviously both rely on this data.

**Which issue(s) this PR fixes**:
Fixes #22702

**Special notes for your reviewer**:
~~Tests will still be added.~~ DONE

s-h-a-d-o-w committed:
Prometheus: Add off switch for metric/label name lookup

This will help users with amounts of metric name
data that is too much for a browser to handle.

Autocomplete will be disabled and metrics chooser hidden,
since obviously both rely on this data.

Fixes #22702

aocenas commented on public/app/plugins/datasource/prometheus/datasource.ts:
```diff
@@ -100,8 +100,11 @@ export class PrometheusDatasource extends DataSourceApi<PromQuery, PromOptions>
     this.customQueryParameters = new URLSearchParams(instanceSettings.jsonData.customQueryParameters);
+
+    if (!instanceSettings.jsonData.disableLanguageProvider) {
+      this.languageProvider = new PrometheusLanguageProvider(this);
```
I think we should change the type of `languageProvider` to `languageProvider?` to reflect this. It necessary right now because of not running strict but will ease some fixes later when we do.

aocenas replied:
Another thing is that this will disable both the metric dropdown but also all autocomplete. Autocomplete also suggest static things like functions which I think do not need server request. So question is if want to do it that way (could make sense some people already asked about switch for autocomplete in general) or we want to keep the autocomplete functionality where possible?

s-h-a-d-o-w replied:
About the type change - yes, thanks! I think I'll switch on strict mode while working just to catch stuff like this in the future. ;)

s-h-a-d-o-w replied:
Well, I for one would hope that this is just a temporary solution anyway but @davkal what do you think? Should I invest a bit more time here and not simply cut off the `languageProvider` completely?

davkal replied:
Yes good point! We don't want do disable the autocomplete entirely (e.g., for function names), but only the parts that use the API for metric and label lookups. (The name of the setting label "Disable metric lookup" is accurate.)

aocenas commented on public/app/plugins/datasource/prometheus/components/PromQueryField.tsx:
```diff
@@ -302,11 +302,13 @@ class PromQueryField extends React.PureComponent<PromQueryFieldProps, PromQueryF
-              {chooserText}
-            </ButtonCascader>
-          </div>
+          {this.languageProvider && (
```
Also lets make the type of `this.languageProvider` property optional here.

s-h-a-d-o-w replied:
What do you mean by this? It **is** optional and that's why the ButtonCascader will render or not, depending on whether it is set?

aocenas replied:
The same thing as in datasource, it is defined as `languageProvider: PromQlLanguageProvider;` and I guess it should be `languageProvider?: PromQlLanguageProvider;`

s-h-a-d-o-w replied:
Ohh, so the same thing that you also noted above? Yeah, that's already addressed. :)

s-h-a-d-o-w commented:
Tests got added. I realized that the only thing we want to protect from here is people changing either `languageProvider` or the conditional in unintended ways within `PromQueryField`, so I went for a significantly "dumber" mock than I originally intended. :wink:

torkelo commented on public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx:
```diff
@@ -130,3 +138,15 @@ const onChangeHandler = (key: keyof PromOptions, value: Props['value'], onChange
   });
 };
+
+const onChangeHandlerSwitch = (key: keyof PromOptions, value: Props['value'], onChange: Props['onChange']) => (
+  eventItem: SyntheticEvent<HTMLInputElement>
+) => {
+  onChange({
+    ...value,
+    jsonData: {
+      ...value.jsonData,
+      [key]: eventItem.currentTarget.checked,
+    },
+  });
+};
```
We already have a function for this in grafana/data , onUpdateDatasourceJsonDataOptionChecked

torkelo commented on public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx:
```diff
@@ -75,11 +75,19 @@ export const PromSettings = (props: Props) => {
       <div className="gf-form-group">
+        <div className="gf-form">
+          <Switch
+            label="Disable query editor metric/label name lookup"
```
this is a bit long for a Switch label for the old form styles (might work after switch to new form style). Maybe just `Disable metric lookup` and rest in the tooltip?

s-h-a-d-o-w committed:
Use onUpdateDatasourceJsonDataOptionChecked

... from `@grafana/data`. Refactor naming to faciliate its
use and stick with prop names as passed down from
`ConfigEditor`.

PLUS:
- Rephrase switch label, add a tooltip and reduce the
size of the to what "Custom query parameters" originally
was.
- Change `languageProvider` type in `PromQueryField`.

davkal commented on public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx:
```diff
@@ -67,14 +68,23 @@ export const PromSettings = (props: Props) => {
+            label="Disable metric lookup"
+            labelClass="width-14"
+            onChange={onUpdateDatasourceJsonDataOptionChecked(props, 'disableLanguageProvider')}
+            tooltip="Switch on if you experience performance issues in the query editor. This will disable autocomplete and the metrics chooser."
```
Should be clarified to only disable the metric and label value lookups.

s-h-a-d-o-w replied:
But how can the metrics chooser work without metrics data? (Might be good if you suggested a concrete wording here because I'm not sure whether I understood your comment correctly.)

davkal replied:
Yeah, we need to make clear that the metrics chooser will be off/disabled/not working if the user enables this option.

davkal replied:
"Checking this option will disable the Metrics chooser and metric/label support in the query field's tab-completion. This helps if you have performance issues with bigger Prometheus instances."

davkal reviewed (changes requested):
As @aocenas already suggested, we'd like to keep static autocomplete features intact.
Other parts of the PR already look good.

s-h-a-d-o-w committed:
Put language provider back in

Functions and history still work, even when metrics
lookup gets disabled.
Also: Rewording of setting.

s-h-a-d-o-w commented:
Alright, I believe I've addressed all of the notes. :)

aocenas commented on public/app/plugins/datasource/prometheus/components/PromQueryField.tsx:
```diff
@@ -270,7 +268,7 @@ class PromQueryField extends React.PureComponent<PromQueryFieldProps, PromQueryF
     // Hint for big disabled lookups
     let hint: QueryHint;
-    if (lookupsDisabled) {
+    if (!datasource.lookupsDisabled && languageProvider.lookupsDisabled) {
```
So do we want to keep both automatic disabling and manual disabling at the same time? Also where do we show this hint? I feel it may be useful to have something like that also for the manual one. Like showing `Dynamic label lookup is disabled in data source config.` so users are less confused why it does not work.

s-h-a-d-o-w replied:
It's not really "at the same time" but rather "possibly different configurations for multiple prometheus data sources". Pretty much in accordance with your suggestion to have a different error message.
I'll break this up and show different messages, good idea.

aocenas replied:
But do I understand correctly that if data source has the lookupsDisabled == false it can still get disabled automatically if we see that the data source has lots of metrics? I don't have a strong opinion here just thought going for manual on/off switch we will remove the automatic one because it wasn't working properly anyway (otherwise we would not need the manual switch).

s-h-a-d-o-w replied:
Well, the way I figure it, the dynamic one might catch a couple of cases where people have >10K metric names but not so many that it makes grafana unusable. Because IIRC, it only gets problematic when they're in the hundreds of thousands to millions. So the automatic disabling may work for people who have something like 10K-200K metric names.

s-h-a-d-o-w committed:
Display a message when lookup got disabled manually

davkal commented on public/app/plugins/datasource/prometheus/types.ts:
```diff
@@ -20,6 +20,7 @@ export interface PromOptions extends DataSourceJsonData {
   httpMethod: string;
   directUrl: string;
   customQueryParameters?: string;
+  disableLanguageProvider?: boolean;
```
can we rename this still before merge?

s-h-a-d-o-w replied:
OH! Great catch, will do!

s-h-a-d-o-w replied:
Done.

s-h-a-d-o-w committed:
Call property for setting disableMetricsLookup

davkal commented:
One final nit (only ran it just now so i saw how it looked): the warning could be in the way of everyday use, i.e., users could think that something is wrong and ask their admins for explanation. 
For this PR i'd rather show nothing, (or still show the metrics chooser disabled with a tooltip of the warning).

s-h-a-d-o-w committed:
Show disabled metrics chooser instead of warning

s-h-a-d-o-w commented:
@davkal Should be good to go now. :)

teu commented:
Hi.

We add a lot of prometheus data sources (one per client) so we have autocomplete. The problem is that the aformentioned setting is not visible once the data source is added. Would it be possible to disable the lookup just for the import bit ?
